operations/mimir/memcached.libsonnet

```diff
@@ -17,26 +19,52 @@ memcached {
       (if !std.isObject($._config.node_selector) then {} else statefulSet.mixin.spec.template.spec.withNodeSelectorMixin($._config.node_selector)) +
       $.util.antiAffinity,
 
-    local service = $.core.v1.service,
-
     service:
       $.util.serviceFor(self.statefulSet) +
       service.mixin.spec.withClusterIp('None'),
   },
 
-  // Dedicated memcached instance used to cache query results.
-  memcached_frontend: $.memcached {
-    name: 'memcached-frontend',
-    max_item_size: '5m',
+  // Optionally configure Memcached to use mTLS for authenticating clients
+  mtls:: {
+    // CA and server credentials
+    statefulSet+: statefulSet.mixin.spec.template.spec.withVolumesMixin([
+      volume.fromSecret($._config.memcached_mtls_ca_cert_secret, $._config.memcached_mtls_ca_cert_secret),
+      volume.fromSecret($._config.memcached_mtls_server_cert_secret, $._config.memcached_mtls_server_cert_secret),
+      volume.fromSecret($._config.memcached_mtls_server_key_secret, $._config.memcached_mtls_server_key_secret),
+    ]),
+
+    // Increase requested CPU, mount the credentials, and adjust server config
+    memcached_container+: $.util.resourcesRequestsMixin('1', null) + container.withVolumeMountsMixin([
+      mount.new($._config.memcached_mtls_ca_cert_secret, $._config.memcached_ca_cert_path, true),
+      mount.new($._config.memcached_mtls_server_cert_secret, $._config.memcached_server_cert_path, true),
+      mount.new($._config.memcached_mtls_server_key_secret, $._config.memcached_server_key_path, true),
+    ]) + container.withArgsMixin([
+      '--listen=notls:127.0.0.1:11211,0.0.0.0:11212', // No TLS on the local interface for the exporter and debugging
+      '--enable-ssl',
+      '--extended=' +
+      'ssl_ca_cert=' + $._config.memcached_ca_cert_path + $._config.memcached_mtls_ca_cert_secret + '.pem,' +
+      'ssl_chain_cert=' + $._config.memcached_server_cert_path + $._config.memcached_mtls_server_cert_secret + '.pem,' +
+      'ssl_key=' + $._config.memcached_server_key_path + $._config.memcached_mtls_server_key_secret + '.pem,' +
+      'ssl_kernel_tls,ssl_verify_mode=2', // "2" means "require client cert"
+    ]),
   },
 
+  // Dedicated memcached instance used to cache query results.
+  memcached_frontend: if $._config.memcached_frontend_enabled then
+    $.memcached {
+      name: 'memcached-frontend',
+      max_item_size: '%dm' % [$._config.memcached_frontend_max_item_size_mb],
+      //connection_limit: 16384,
```

Comment:
Should be set for consistency but this would trigger a rollout of every instance.

Reply:
Same comment here. I think it's relatively easy to set it here, and then revert in GL until next release.